File: Sources/AINotebookCore/TransformationEngine.swift
import Foundation

public actor TransformationEngine {
    public enum RunError: Error, Equatable {
        case sourceNotFound(Int64)
        case transformationNotFound(Int64)
        case noChunks(Int64)
    }

    private let store: NotebookStore
    private let chat: ChatStreaming
    public let chatModel: String

    public init(store: NotebookStore, chat: ChatStreaming, chatModel: String) {
        self.store = store
        self.chat = chat
        self.chatModel = chatModel
    }

    @discardableResult
    public func run(
        transformationId: Int64,
        sourceId: Int64,
        onToken: @escaping @Sendable (String) -> Void = { _ in }
    ) async throws -> Note {
        let storeRef = store

        let prep: (Transformation, Source, [SourceChunk]) =
            try await MainActor.run {
                guard let t = try storeRef.transformations().first(where: { $0.id == transformationId }) else {
                    throw RunError.transformationNotFound(transformationId)
                }
                guard let s = try storeRef.source(id: sourceId) else {
                    throw RunError.sourceNotFound(sourceId)
                }
                let cs = try storeRef.chunks(sourceId: sourceId)
                return (t, s, cs)
            }
        let (transformation, source, chunks) = prep
        guard !chunks.isEmpty else { throw RunError.noChunks(sourceId) }

        let sourceText = chunks.map(\.text).joined(separator: "\n\n")
        let rendered = transformation.promptTemplate
            .replacingOccurrences(of: "{{source_text}}", with: sourceText)

        let turns: [ChatTurn] = [
            ChatTurn(role: .user, content: rendered)
        ]
        var assembled = ""
        for try await token in chat.stream(model: chatModel, messages: turns) {
            assembled += token
            onToken(token)
        }

        let noteTitle = "\(transformation.name) — \(source.title)"
        let note = try await MainActor.run {
            let created = try storeRef.createNote(
                notebookId: source.notebookId,
                title: noteTitle,
                bodyMd: assembled,
                origin: .transformation,
                originRef: transformation.id
            )
            _ = try storeRef.recordTransformationRun(
                transformationId: transformation.id!,
                sourceId: source.id!,
                resultNoteId: created.id
            )
            return created
        }
        return note
    }

    @discardableResult
    public func runNotebookScope(
        transformationId: Int64,
        notebookId: Int64,
        onToken: @escaping @Sendable (String) -> Void = { _ in }
    ) async throws -> Note {
        let storeRef = store
        let prep: (Transformation, [Source], [SourceChunk]) =
            try await MainActor.run {
                guard let t = try storeRef.transformations().first(where: { $0.id == transformationId }) else {
                    throw RunError.transformationNotFound(transformationId)
                }
                let sources = try storeRef.sources(notebookId: notebookId)
                var allChunks: [SourceChunk] = []
                for s in sources {
                    allChunks.append(contentsOf: try storeRef.chunks(sourceId: s.id!))
                }
                return (t, sources, allChunks)
            }
        let (transformation, sources, chunks) = prep
        guard !chunks.isEmpty else { throw RunError.noChunks(notebookId) }

        let sourceText = chunks.map(\.text).joined(separator: "\n\n")
        let rendered = transformation.promptTemplate
            .replacingOccurrences(of: "{{source_text}}", with: sourceText)

        let turns: [ChatTurn] = [ChatTurn(role: .user, content: rendered)]
        var assembled = ""
        for try await token in chat.stream(model: chatModel, messages: turns) {
            assembled += token
            onToken(token)
        }
        let noteTitle = "\(transformation.name) — \(sources.count) sources"
        let note = try await MainActor.run {
            let created = try storeRef.createNote(
                notebookId: notebookId,
                title: noteTitle,
                bodyMd: assembled,
                origin: .transformation,
                originRef: transformation.id
            )
            _ = try storeRef.recordTransformationRun(
                transformationId: transformation.id!,
                sourceId: nil,
                resultNoteId: created.id
            )
            return created
        }
        return note
    }
}

File: windows/src/AINotebook.Core/Rag/TransformationEngine.cs
using AINotebook.Core.Models;
using AINotebook.Core.Ollama;
using AINotebook.Core.Storage;

namespace AINotebook.Core.Rag;

public sealed class TransformationEngine
{
    private readonly NotebookStore _store;
    private readonly IChatStreaming _chat;
    public string ChatModel { get; }

    public TransformationEngine(NotebookStore store, IChatStreaming chat, string chatModel)
    {
        _store = store;
        _chat = chat;
        ChatModel = chatModel;
    }

    public async Task<Note> RunAsync(
        long transformationId, long sourceId, Action<string>? onToken = null, CancellationToken ct = default)
    {
        var transformation = _store.Transformations().FirstOrDefault(t => t.Id == transformationId)
            ?? throw new TransformationException.TransformationNotFound(transformationId);
        var source = _store.Source(sourceId)
            ?? throw new TransformationException.SourceNotFound(sourceId);
        var chunks = _store.Chunks(sourceId);
        if (chunks.Count == 0) throw new TransformationException.NoChunks(sourceId);

        var sourceText = string.Join("\n\n", chunks.Select(c => c.Text));
        var rendered = transformation.PromptTemplate.Replace("{{source_text}}", sourceText);

        var assembled = await StreamAssembleAsync(rendered, onToken, ct);

        // em-dash U+2014 in the title separator.
        var noteTitle = $"{transformation.Name} — {source.Title}";
        var created = _store.CreateNote(source.NotebookId, noteTitle, assembled,
            NoteOrigin.Transformation, transformation.Id);
        _store.RecordTransformationRun(transformation.Id!.Value, source.Id!.Value, created.Id);
        return created;
    }

    public async Task<Note> RunNotebookScopeAsync(
        long transformationId, long notebookId, Action<string>? onToken = null, CancellationToken ct = default)
    {
        var transformation = _store.Transformations().FirstOrDefault(t => t.Id == transformationId)
            ?? throw new TransformationException.TransformationNotFound(transformationId);
        var sources = _store.Sources(notebookId);
        var allChunks = new List<SourceChunk>();
        foreach (var s in sources) allChunks.AddRange(_store.Chunks(s.Id!.Value));
        if (allChunks.Count == 0) throw new TransformationException.NoChunks(notebookId);

        var sourceText = string.Join("\n\n", allChunks.Select(c => c.Text));
        var rendered = transformation.PromptTemplate.Replace("{{source_text}}", sourceText);

        var assembled = await StreamAssembleAsync(rendered, onToken, ct);

        var noteTitle = $"{transformation.Name} — {sources.Count} sources";
        var created = _store.CreateNote(notebookId, noteTitle, assembled,
            NoteOrigin.Transformation, transformation.Id);
        _store.RecordTransformationRun(transformation.Id!.Value, null, created.Id);
        return created;
    }

    public async Task<IReadOnlyList<Note>> RunOnAllSourcesAsync(
        long transformationId, long notebookId, Action<int, int>? onProgress = null, CancellationToken ct = default)
    {
        var sources = _store.Sources(notebookId);
        var total = sources.Count;
        var results = new List<Note>();
        for (var idx = 0; idx < sources.Count; idx++)
        {
            var note = await RunAsync(transformationId, sources[idx].Id!.Value, null, ct);
            results.Add(note);
            onProgress?.Invoke(idx + 1, total);
        }
        return results;
    }

    private async Task<string> StreamAssembleAsync(string rendered, Action<string>? onToken, CancellationToken ct)
    {
        var turns = new List<ChatTurn> { new(ChatRole.User, rendered) }; // NO system turn
        var assembled = "";
        await foreach (var token in _chat.StreamAsync(ChatModel, turns, ct))
        {
            assembled += token;
            onToken?.Invoke(token);
        }
        return assembled;
    }
}
